RFC: Establish an RFC process

cmnrd opened the pull request:
[rendered](https://github.com/lf-lang/rfcs/blob/rfc-process/rfcs/0002-rfc-process.md)

This is a proposal to introduce a request for comments (RFC) process in the LF community. To bootstrap this RFC process, this proposal follows the proposed RFC style. Please read and review this RFC carefully. Any feedback or comments are highly appreciated. The main goal of this proposal is to establish a process for introducing new features into the LF language and tools.

tanneberger commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+    request, and leave a comment on the pull request explaining your changes.
+    Specifically, do not squash or rebase commits after they are visible on the
+    pull request.
+  - At some point, a member of the team will propose a "motion for final
```
FYI: in IETF speak this would be called "WG Last Call"

lhstrh replied:
I think "motion" is clearer than "last call." (As used in Robert's Rules, for instance.)

tanneberger commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+  - At some point, a member of the team will propose a "motion for final
+    comment period" (FCP), along with a *disposition* for the RFC (merge, close,
+    or postpone).
+    - This step is taken when enough of the tradeoffs have been discussed that
```
The proper term would be : "rough consensus" 
-> https://datatracker.ietf.org/doc/html/rfc7282

tanneberger commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+## The RFC life-cycle
+[The RFC life-cycle]: #the-rfc-life-cycle
+
+Once an RFC becomes "active" then authors may implement it and submit the
```
I think it is very desirable to have an implementation ready before the Draft is merged. 

-> Normally, the code and the rfc are developed side by side, this gives the benefit that the rfc is enriched by developer / implementer feedback

cmnrd replied:
Maybe the sentence is formulated a bit too strict here. Of course, a contributor may also work on an implementation earlier on. It would be risky, however. The main goal of RFCs as proposed here is to build an agreement behind a proposal, not behind an implementation. Also, it acts as a mechanism for contributors to get feedback early on so that they can understand if the feature they are planning is desirable before hey put in the work of implementing it. I think this goal is likely different from the IETF RFCs.

tanneberger replied:
I think developing them side by side is a good feasibility test and avoids situations where proposals look good on paper but not in reality.

cmnrd replied:
That would be a major change to the proposal though. The proposed process explicitly focuses on discussing ideas and features, not implementations. The main idea is to commit to thinking an idea through before implementing it. That does not rule out that an RFC is guided by an implementation, but I think it shouldn't be the norm.

Of course it can be complicated or infeasible to implement an RFC. If it is unclear how the RFC could be implemented, this should be part of the discussion. It is also possible to list implementation aspects under "Unresolved Questions" in the proposed template.

Also note that the proposed process explicitly states that active RFCs are not a guarantee that an implementation PR actually gets merged. If the implementation is bad the PR should be rejected. If the implementation is bad because of a flaw in the RFC, then the RFC should be amended or replaced.

The proposed RFC process is intended to be rather lightweight. If we include also the implementation, it will be a significantly longer process.

erlingrj replied:
I think I agree with Christian here. I think our current workflow, with thorough code reviews and CI, works quite well for guiding the development of features. But what lacks is a structured way of agreeing on what features we should target (which is exactly what the RFC process is about).

An example is the work I am doing on the [physical clock abstraction in the C runtime]( https://github.com/lf-lang/reactor-c/pull/346), the question of how the LF runtime system should react if a clock synchronization client running on the system, like `chrony` or `ptpd` steps the clocks back or forward 100ms is not only a problem for the C target. I think such questions would be great to discuss through an RFC at the "idea-level"

petervdonovan replied:
> Of course, a contributor may also work on an implementation earlier on. It would be risky, however.

I agree with this. I think it is important for anyone who works on an implementation before the RFC is approved to be aware that there is a significant risk that their work will be discarded.

> I think such questions would be great to discuss through an RFC at the "idea-level"

I also agree with this, I think that we should have high-level discussions that have the potential to change or reject a proposed feature and that don't get bogged down in details about the implementation.

> I think developing them side by side is a good feasibility test and avoids situations where proposals look good on paper but not in reality.

But on the other hand I also do agree with this. I do not know about anyone else, but if I were to propose a feature, I might be much more comfortable discussing it if I already had a prototype or partial implementation. I think it makes sense to tolerate either approach (implement before RFC approval, or wait for approval) as long as everyone understands the risk of rejecting a premature implementation and as long as people agree to focus RFCs on high-level discussion

tanneberger commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+## Implementing an RFC
+[Implementing an RFC]: #implementing-an-rfc
+
+Some accepted RFCs represent vital features that need to be implemented right
```
I would like to raise the point about what is officially supported. 

-> What are the sections of code that always should have feature parity?

cmnrd replied:
Could you elaborate on what you mean here? I don't understand the question, unfortunately.

tanneberger replied:
Okay, let us say we approve a new RFC that adds some runtime/language features. The question is, what are significant changes and what are not? I would argue that we should define a language core and that any RFC that changes the language core should be merged in one go. I think we'd be shooting ourselves in the foot if a Lingua Franca release has only a half-functioning feature (e.g. C target has support, but CPP doesn't).

lhstrh replied:
I think that this might be too tall of an order. It rules out features that cannot be supported by all targets and it could cause features to be delayed indefinitely if maintainers of some targets prioritize other things...

cmnrd replied:
Its also an orthogonal question in my opinion. The proposed RFC process explicitly focuses on features, not implementations. Accepted (active) RFCs are an indication that the community is interested in a specific feature. It does not mean that there is or will be a complete implementation of that feature in the near term. It could also be that an active RFC is never implemented because it simply isn't a priority.

The point you raise is important, too. How do we manage the heterogeneity of features between the LF targets? How do we ensure some level of feature parity? These are beyond the scope of the RFC process as proposed here. I also don't see a clear way for including these questions in our process. Do you have a concrete proposal? Otherwise, I would say this is a great topic for a future RFC.

tanneberger commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+## License
+[License]: #license
+
+TODO
```
See my comments on martens issue #1

tanneberger commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+- Tracking Issue: [lf-lang/lingua-franca#0000](https://github.com/lf-lang/lingua-franca/issues/0000)
+
+# Summary
+[summary]: #summary
```
abstract

cmnrd replied:
Is this a proposed change? Is there any reasoning behind it?

tanneberger replied:
yeah I prefer abstract to summary.

lhstrh replied:
~~If it's an abstract it should appear at the beginning of the document, so I think summary is adequate here...~~ Sorry, I didn't realize this was the beginning of the template :man_facepalming:  Yes, "abstract" should work here.

cmnrd replied:
Alright, we can rename this section.

Yes, the idea of the RFC template is that it includes the documentation of the proposed feature. Which in this case is the template and the README.

tanneberger commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+`0000-template.md` to the root of the repository to act as a template for new
+RFCs.
+
+# Drawbacks
```
Normally there are sections called like: `Security Considerations`, `Deployment Recommendations` 

-> Where the potential impact on the ecosystem or users is described.

cmnrd replied:
I think those terms do not apply in our case. Which considerations are relevant to an RFC depends on the precise scope of the RFC. Therefore, I would argue we should keep it general. If we notice that certain points are always requested, we can still update the template.

edwardalee commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+
+- Which license should we use for RFCs in this repository?
+- How can we send out FCPs to all interested parties? (The mailing list would
+  work, put currently there is no way for people to subscribe)
```
```suggestion
  work, but currently there is no way for people to subscribe)
```

edwardalee reviewed (approved):
This looks good to me, although I worry it will create extra work for the "core team." In exchange for that extra work, changes will be better, better documented, and there will be stronger consensus behind them.  But keep in mind that extra work inevitably means less getting done on the "to do" list, so there is a cost. As the community gets larger, a process like this will be essential.  But is now the time?

cmnrd commented:
When would there be a better time? I mention under "Drawbacks" that this process may slow down development. This is not necessarily a bad thing, however. I would argue that slow development in the right direction (as defined by the community) is better than fast development (potentially) in the wrong direction.

I would be very open for proposals on how we can keep the overhead low or also on alternative processes. Continuing as we did so far, however, does not seem to prepare us well for the upcoming challenges. I would argue that our community is already large enough and driven by different interests, so that different sub groups move in different directions. We can now try to reunite the different directions and formulate a united way forward, which might come at the expense of overall velocity, or we can let the fastest developers dominate the direction the project takes, which would likely lead to a split community. Both are viable options, but the outcome will be significantly different.

edwardalee commented:
Yes, on balance, I think this is a good proposal and am willing to move forward with it.

lhstrh commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+## The RFC life-cycle
+[The RFC life-cycle]: #the-rfc-life-cycle
+
+Once an RFC becomes "active" then authors may implement it and submit the
```
```suggestion
Once an RFC becomes "active," anyone may submit an implementation of the
```

lhstrh commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+community. However, for now, we should try to establish a process that works for
+us (and refine it) before automating the process.
+
+- We can automatically generate a website that lists all active RFCs. The Rust
```
I'm in favor of using `mdbook` for this purpose as well.

lhstrh commented:
This looks great! Please run a formatter on the Markdown before merging :rocket:

petervdonovan commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+
+A variation of the proposed RFC process has been used successfully by the Rust
+community for the last 10 years. The process is intended to enforce a more
+disciplined approach wile being as light-weight as possible. The main rationale
```
```suggestion
disciplined approach while being as light-weight as possible. The main rationale
```

petervdonovan commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+If there is no prior art, that is fine - your ideas are interesting to us whether they are brand new or if it is an adaptation from solutions.
+
+Note that while precedent set by other languages is some motivation, it does not on its own motivate an RFC.
+Please also take into consideration that rust sometimes intentionally diverges from common language features.
```
This comment about Rust may not apply to LF

lhstrh replied:
If you change "rust" to "Lingua Franca" the sentence sounds reasonable to me though!

petervdonovan commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+- For other teams: What lessons can we learn from what other communities have done here?
+- Papers: Are there any published papers or great posts that discuss this? If you have some relevant papers to refer to, this can serve as a more detailed theoretical background.
+
+This section is intended to encourage you as an author to think about the lessons from other languages and frameworks, provide readers of your RFC with a fuller picture.
```
```suggestion
This section is intended to encourage you as an author to think about the lessons from other languages and frameworks, and to provide readers of your RFC with a fuller picture.
```

edwardalee commented:
Most ideas that I work on cannot be thoroughly thought through before implementing them.  It sounds like you are advocating a waterfall style of software engineering, where a specification is fully fleshed out before any code is written.  I think this style of software engineering has very much fallen out of favor.  It has certainly never worked for me.

lhstrh commented on text/0001-rfc-process.md:
```diff
@@ -0,0 +1,428 @@
+If there is no prior art, that is fine - your ideas are interesting to us whether they are brand new or if it is an adaptation from solutions.
+
+Note that while precedent set by other languages is some motivation, it does not on its own motivate an RFC.
+Please also take into consideration that rust sometimes intentionally diverges from common language features.
```
```suggestion
Please also take into consideration that Lingua Franca sometimes intentionally diverges from common language features.
```

lhstrh commented:
> Most ideas that I work on cannot be thoroughly thought through before implementing them. It sounds like you are advocating a waterfall style of software engineering, where a specification is fully fleshed out before any code is written. I think this style of software engineering has very much fallen out of favor. It has certainly never worked for me.

I personally don't read the proposal that way. To me, the RFC is a method to vet a proposal, and the process of handling RFCs and their implementation is meant to ensure that only well-vetted proposal make it into the core tool chain. It is entirely up to the individual contributors whether they think it is worth the risk of fleshing out an implementation before validating the idea and without any guarantee of getting it merged.

cmnrd commented:
> Most ideas that I work on cannot be thoroughly thought through before implementing them. It sounds like you are advocating a waterfall style of software engineering, where a specification is fully fleshed out before any code is written. I think this style of software engineering has very much fallen out of favor. It has certainly never worked for me.

There might be a bit of a misunderstanding here. The scope of the RFC process is not the entire development cycle from specification to implementation. It is really just a small part of the cycle. How fine-grained we make the cycle is up to us (and probably also differs case by case).

RFCs (as proposed here) are explicitly not a full specification and, most importantly, never final. The intent is to establish a process for design proposals. Accepted (active) RFCs are a collection of ideas that the community considers both relevant and feasible. As requirements and the community's goals change, the active RFCs should also change. And if we realize during implementation that a design is not feasible or that the RFCs did not consider important aspects of the implementation, then the RFC should be amended or replaced by a new RFC. I see this as something very organic. In a sense, the active RFCs are a snapshot of the community's goals and understanding of the project. 

I think we all agree that we should not aim for a waterfall model. As we implement features, requirements and our understanding of the design change. However, it is important nonetheless to set a direction and to inform others about the next steps one plans to take, especially if those steps potentially impact others. An accepted RFC is not a commitment to take precisely those steps. It is completely fine (and encouraged!) to adjust the path along the way. A goal of this proposal is to make this process more visible and involve others in the relevant decisions.

Quote from the proposal:
> Modifications to "active" RFCs can be done in follow-up pull requests. We strive to write each RFC in a manner that it will reflect the final design of the feature; but the nature of the process means that we cannot expect every merged RFC to actually reflect what the end result will be at the time of the next major release.


Likely, we will need to adjust our initial RFCs more frequently. But I suspect that we will get better over time. Also, I expect our initial RFCs to focus on bigger questions, which quite naturally will require more iterations. But once we have settled on the foundation, RFCs will likely become a lot more fine-grained. Many of the Rust RFCs, for instance, just consider one specific behavior of a specific command line argument.

I would also like to note that the only rule with respect to implementation is that there should be an accepted RFC before a major feature PR gets accepted. If the RFC and the implementation are developed side by side, this is completely fine. In fact, reviewers could also request such a prototype if they have doubts about the feasibility of an idea. The general intent, however, is to first file an RFC and get feedback on the idea before investing in a concrete implementation. But this is up to the author to decide.

cmnrd commented:
I included the given feedback in https://github.com/lf-lang/rfcs/pull/2/commits/0d255fe1fdf4cedadb74a1035978100517db9bb4. Please check.

Summary of changes:
 - Rename Summary to Abstract
- Clarify the role of RFCs in the design cycle
- Include the term "rough consensus"
- Acknowledge that the proposed process bears the risk of being interpreted as a specification and implemented in a waterfall model
- Acknowledge that any design review is incomplete without an implementation
- Explain the rationale behind focusing on the design and not a concrete implementation

soerendomroes commented:
I like the idea to use this process for proposals. Our group usually does something similar for each thesis.

I do, however, think that the documentation could be shorter (I try to find time for more specific comments tomorrow) and I think there could be something between an RFC and a PR (maybe a discussion). But maybe time will show that a discussion is just a small RFC.

cmnrd commented:
Based on our discussion in the weekly LF meeting, I made the following additions on when RFCs should be used:
 - Changes to the graphical syntax of Lingua Franca.
 - Changes that require significant updates of the handbook.

and when they should not be used:
- Trivial changes that can be implemented and discussed in a single PR.

In the meeting, Shaokai raised the question if the Rust process is really a good template considering that our team is much smaller. [This article](https://blog.pragmaticengineer.com/scaling-engineering-teams-via-writing-things-down-rfcs/) summarizes perfectly how and why RFCs work and also explicitly states that RFCs are recommended for small teams, in particular when they plan to scale up. I also added the link under prior art.

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+  - Changes to the interfaces between other components of the LF ecosystem, e.g., the language server.
+  - New user facing features in the IDEs and related tools.
+  - Changes to the graphical syntax of Lingua Franca.
+  - Changes thas require signifcant updates of the handbook.
```
signifcant -> significant

petervdonovan replied:
```suggestion
  - Changes that require significant updates of the handbook.
```

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+    be poorly-received.
+  - Submit a pull request (the PR title should start with "RFC:"). As a pull
+    request the RFC will receive design feedback from the larger community, and
+    the author should be prepared to revise it in response.
```
I think the important part is that in here is that the PR should have a title beginning with "RFC:". I would not hide this information in brackets (I did not find it for my PR). Maybe:
Submit a pull request titled "RFC: <your feature>

Everything after this is information that might be helpful but since it is part of the same paragraph, it makes the really important information (submit and name the PR) less visible.

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+    request the RFC will receive design feedback from the larger community, and
+    the author should be prepared to revise it in response.
+  - Now that your RFC has an open pull request, use the issue number of the PR
+    to update your `0000-` prefix to that number.
```
Maybe only:
Update your `0000-` prefix to your PR number.

The rest seems to be a filler.

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+  - Build consensus and integrate feedback. RFCs that have broad support are
+    much more likely to make progress than those that don't receive any
+    comments. Feel free to reach out in particular to get help identifying
+    stakeholders and obstacles.
```
It is not clear to me how this will work, and it seems to be overlapping with the next point:
Do we need this as a separate step?

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+    clarify or change the design, but make changes as new commits to the pull
+    request, and leave a comment on the pull request explaining your changes.
+    Specifically, do not squash or rebase commits after they are visible on the
+    pull request.
```
I think this is interesting but it does not fit into the RFC recipe above.

Maybe just "Do not squash or rebase commits in the RFC PR" is enough or move this to the front and add a linebreak after it for details?

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+      a *summary comment* trying to lay out the current state of the discussion
+      and major tradeoffs/points of disagreement.
+    - The motion for FCP should include a checkbox list that tags major
+      stakeholder and reviewers that need to sign off on the FCP. This is often
```
This sentence is very important for the process and should be more visible. Everything after this seems to be significant but more of an explanation.

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+    days. It is also advertised widely,
+    e.g. on the LF mailing list. This way all
+    stakeholders have a chance to lodge any final objections before a decision
+    is reached.
```
Maybe add this information to the "At some point, a member" FCP part above to have all relevant FCP information in one point.

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+    is reached.
+  - In most cases, the FCP period is quiet, and the RFC is either merged or
+    closed. However, sometimes substantial new arguments or ideas are raised,
+    the FCP is canceled, and the RFC goes back into development mode.
```
This seems like something that is nice to know, but is not essential to the process.

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+implementing the feature. While it is not *necessary* that the author of the
+RFC also write the implementation, it is by far the most effective way to see
+an RFC through to completion: authors should not expect that other project
+developers will take on responsibility for implementing their accepted feature.
```
I think
"Once an RFC becomes "active," anyone may submit an implementation of the
feature as a pull request to the relevant repo."
is the important information here. The rest is an explanation of what this means. Do we want to somehow highlight that?

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+strive to write each RFC in a manner that it will reflect the final design of
+the feature; but the nature of the process means that we cannot expect every
+merged RFC to actually reflect what the end result will be at the time of the
+next release.
```
Again: "Modifications to "active" RFCs can be done in follow-up pull requests." is the important information that gets mixed with the explanation of what this means.

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+In general, once accepted, RFCs should not be substantially changed. Only very
+minor changes should be submitted as amendments. More substantial changes
+should be new RFCs, with a note added to the original RFC. Exactly what counts
+as a "very minor change" is up to the team to decide.
```
This should be part of the previous paragraph. 
Maybe:
"Minor modifications to "active" RFCs can be done in follow-up PRs. Substantial changes require a new RFC."

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+  - Removing language features.
+  - Changes to the interface between the compiler and the runtime libraries.
+  - Changes to the interfaces between other components of the LF ecosystem, e.g., the language server.
+  - New user facing features in the IDEs and related tools.
```
```suggestion
  - New user-facing features in the IDEs and related tools.
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+
+Some changes do not require an RFC:
+
+  - Rephrasing, reorganizing, refactoring, or otherwise "changing shape does
```
```suggestion
  - Rephrasing, reorganizing, refactoring, or otherwise "changing shape that does
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+## Before creating an RFC
+[Before creating an RFC]: #before-creating-an-rfc
+
+A hastily-proposed RFC can hurt its chances of acceptance. Low quality
```
```suggestion
A hastily-proposed RFC can hurt its chances of acceptance. Low-quality
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+
+A hastily-proposed RFC can hurt its chances of acceptance. Low quality
+proposals, proposals for previously-rejected features, or those that don't fit
+into the near-term roadmap, may be quickly rejected, which can be demotivating
```
```suggestion
proposals, proposals for previously-rejected features, or proposals that don't fit
into the near-term roadmap may be quickly rejected, which can be demotivating
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+
+Although there is no single way to prepare for submitting an RFC, it is
+generally a good idea to pursue feedback from other project developers
+beforehand, to ascertain that the RFC may be desirable; having a consistent
```
```suggestion
beforehand to ascertain that the RFC may be desirable. Having a consistent
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+[What the process is]: #what-the-process-is
+
+In short, to get a major feature added to Lingua Franca, one must first get the
+RFC merged into the RFC repository as a markdown file. At that point the RFC is
```
```suggestion
RFC merged into the RFC repository as a Markdown file. At that point the RFC is
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+  - Fork the RFC repo [RFC repository]
+  - Copy `0000-template.md` to `text/0000-my-feature.md` (where "my-feature" is
+    descriptive). Don't assign an RFC number yet; This is going to be the PR
+    number and we'll rename the file accordingly if the RFC is accepted.
```
```suggestion
    descriptive). Don't assign an RFC number yet; this is going to be the PR
    number, and we'll rename the file accordingly if the RFC is accepted.
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+Once an RFC becomes "active," anyone may submit an implementation of the
+feature as a pull request to the relevant repo. Being "active" is not a rubber
+stamp, and in particular still does not mean the feature will ultimately be
+merged; it does mean that in principle all the major stakeholders have agreed
```
```suggestion
stamp, and in particular it still does not mean the feature will ultimately be
merged. However, it does mean that in principle, all the major stakeholders have agreed
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+it imply anything about whether a developer has been assigned the task of
+implementing the feature. While it is not *necessary* that the author of the
+RFC also write the implementation, it is by far the most effective way to see
+an RFC through to completion: authors should not expect that other project
```
```suggestion
an RFC through to completion: Authors should not expect that other project
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+  water-fall style without considering changing requirements and new insights
+  gained from the implementation.
+- RFCs focus on the design, not an implementation. Thus, active RFCs will rarely
+  describe the end-result and need to be adjusted when the feature gets
```
```suggestion
  describe the end result and may often need to be adjusted when the feature gets
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+us (and refine it) before automating the process.
+
+- We can automatically generate a website that lists all active RFCs. The Rust
+community uses mdbook for this.
```
```suggestion
community uses mdBook for this.
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+proposed process will not fit exactly our needs and can be modified in future
+RFCs.
+
+In order to be light-weight, the proposed process only considers the design of a
```
```suggestion
In order to be lightweight, the proposed process only considers the design of a
```

petervdonovan reviewed (approved):
I suggested minor clarifying changes because I anticipate that this document will be read many times.

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+## Implementing an RFC
+[Implementing an RFC]: #implementing-an-rfc
+
+Some accepted RFCs represent vital features that need to be implemented right
+away. Other accepted RFCs can represent features that can wait until some
+arbitrary developer feels like doing the work. Every accepted RFC has an
+associated issue tracking its implementation in the relevant repository.
+
+The author of an RFC is not obligated to implement it. Of course, the RFC
+author (like any other developer) is welcome to post an implementation for
+review after the RFC has been accepted.
+
+If you are interested in working on the implementation for an "active" RFC, but
+cannot determine if someone else is already working on it, feel free to ask
+(e.g. by leaving a comment on the associated issue).
```
Is this really necessary to document this for our RFC workflow?

I think the interesting part for the implementation would be how issues, PRs, or commits might be associated to an RFC, should the RFC number be used for this? If yes how?

cmnrd replied:
Each active RFC should have a tracking issue (that itself may contain smaller issues/tasks). Commits and PRs that implement the feature completely or provide increments towards a complete implementation should reference the tracking issue. (This isn't any different from how we handle it how, so probably we don't need to mention it explicitly)

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+- If applicable, describe the differences between teaching this to existing Lingua Franca programmers and new Lingua Franca programmers.
+- Discuss how this impacts the ability to read, understand, and maintain Lingua Franca code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain?
+
+For implementation-oriented RFCs (e.g. for compiler or runtime internals), this section should focus on how compiler or runtime contributors should think about the change, and give examples of its concrete impact.
```
If I remember the last LF meeting correctly, the main point was that this should be the documentation that is copied in the handbook. Maybe write this instead of:
"Explain the proposal as if it was already included in the language and you were teaching it to another Lingua Franca programmer. That generally means:"

soerendomroes commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,469 @@
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Think about what the natural extension and evolution of your proposal would
+be and how it would affect the language and project as a whole in a holistic
+way. Try to use this section as a tool to more fully consider all possible
+interactions with the project and language in your proposal.
+Also consider how this all fits into the roadmap for the project
+and of the relevant sub-team.
+
+This is also a good place to "dump ideas", if they are out of scope for the
+RFC you are writing but otherwise related.
+
+If you have tried and cannot think of any future possibilities,
+you may simply state that you cannot think of anything.
+
+Note that having something written down in the future-possibilities section
+is not a reason to accept the current or a future RFC; such notes should be
+in the section on motivation or rationale in this or subsequent RFCs.
+The section merely provides additional information.
```
I propose to use shorter bullet points:
- How might the feature evolve in the future?

The rest is very wordy and one has to read everything to find things that might be interesting.

soerendomroes reviewed (changes requested):
I do not understand the linebreaks in the markdown file. Why are explicit linebreaks in there that break sentences apart such that only one word remains on either side? Word-wrap should be supported by any modern editor.

I think this is generally too long. I hope you consider some of my suggestions:

I would like to have a bullet-point guide, one can follow to create an RFC. If it is too long one cannot parse the relevant information.

I would like something like:

1. Create a PR in rfcs by copying the markdown file
...
n. Create a PR for your feature

Everything else is just explanation and should not be part of the guide but maybe below for more information as it is done for "Reviewing RFCs".

cmnrd commented:
Thanks for the suggestions! I will work through them. I agree that it is a bit wordy in some places and condensing it down would be helpful.

> I do not understand the linebreaks in the markdown file. Why are explicit linebreaks in there that break sentences apart such that only one word remains on either side? Word-wrap should be supported by any modern editor.
 
Word-wrap is supported by editors, but not well-supported by git, unfortunately. I like to easily see in a diff where a change actually happened. Without breaking lines, it is nearly impossible to identify the relevant change in a long paragraph. I will use an auto-formatter (and maybe also install a CI check) to fix the weird artifacts.

petervdonovan commented:
> Without breaking lines, it is nearly impossible to identify the relevant change in a long paragraph.

I am totally happy no matter which way we format MarkDown files but just in case this is generally useful to you: `git diff --word-diff=color` can solve this. If the command is too long then you can add it to your Git aliases.

lhstrh commented:
Here is a summary of the text provided by ChatGPT:

> The provided text outlines the establishment of a Request for Comments (RFC) process for the Lingua Franca language, compiler infrastructure, tooling, and runtime libraries. The motivation behind this process is to provide a structured approach for proposing new features, enabling all stakeholders to participate in the design process and ensuring confidence in the project's direction.

> The RFC process involves several stages, including proposing an RFC, discussion, review, implementation, and final decision-making. Notably, substantial changes to Lingua Franca, such as semantic or syntactic alterations, require following this process. The text provides guidelines for creating RFCs, including a template for the RFC document. It also discusses the RFC life-cycle, reviewing RFCs, implementing them, and handling postponement.

> Furthermore, the text acknowledges potential drawbacks and alternatives to the proposed RFC process, drawing inspiration from similar processes in other programming communities like Rust and Python. It also addresses unresolved questions and suggests future possibilities for refining and optimizing the RFC process to suit the needs of the Lingua Franca community as it evolves.

> Overall, the RFC process aims to foster a collaborative and disciplined approach to the development of Lingua Franca, ensuring that proposed changes are thoroughly evaluated and aligned with the project's goals and standards.

To summarize the discussion, there appears to be consensus that the proposed text has the right tone and content, and there is a willingness to participate in the RFC process by the key stakeholders that have chimed in. The remaining unresolved comments are mostly focused on representation (not so much on content), which can be addressed in an editorial pass over the text. 

**I move to let this RFC enter the final comment period (FCP), along with the disposition to merge.**

cmnrd commented:
Not so fast ;) There are still some adjustments I would like to make (based on Sören's comments), and we still have two unresolved questions

cmnrd commented:
After letting some of the feedback sink in and doing more research, I have some doubts about the process. I think it has too strong a focus on consensus building and is a bit too formal. In particular, the process on how we move to the final comment period and who should approve is not well specified. I think we will need something simpler, especially in the beginning. I will push a revised, leaner proposal soon.

Here are some interesting resources on the topic:
- [How and why RFCs fail](https://candost.blog/how-and-why-rfcs-fail/)
- [A thorough team guide to RFCs](https://leaddev.com/technical-decision-making/thorough-team-guide-rfcs)
- [Scaling Engineering Teams via RFCs: Writing Things Down](https://blog.pragmaticengineer.com/scaling-engineering-teams-via-writing-things-down-rfcs/)
- [We need to talk about [Rust] RFCs](https://www.ncameron.org/blog/the-problem-with-rfcs/)

petervdonovan commented:
While you are making a second pass I suppose you could consider making the guide-level explanation optional. It might be just a reflection of the RFC format being originally intended for language proposals, while for other kinds of proposals it might be excessive to require three levels of detail (summary, guide-level explanation, and reference-level explanation).

Or as a more high-level takeaway, it looks like a couple of the sources you link mention that it can be a problem for the template to have too many required sections.

cmnrd commented:
I made a complete reiteration of this proposal. In particular, I explicitly state the goals and non-goals. I also refined the process itself, trying to find a balance between quick and small proposals as well as proposals with a larger impact that involve many stakeholders. I also reduced the size (including the number of sections) of the template considerably.

Please read carefully and let me know what you think.

I already addressed some concerns about specific language that were raised during the LF meeting. However, we still have to resolve how we want to implement the decision process in case there is no clear agreeable solution.

edwardalee commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,485 @@
+
+Follow these steps to open a new RFC:
+1. Fork or clone this repository.
+2. Copy [`0000-template.md`](0000-template.md) to `text/0000-my-feature.md`,
```
Where does one find `0000-template.md`? This link does not work.

cmnrd replied:
It is not there yet. The nature of this RFC is to propose the content of this file. The template starts [here](https://github.com/lf-lang/rfcs/blob/rfc-process/text/0002-rfc-process.md#abstract-1) in the rendered proposal.

lhstrh commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,485 @@
+[unresolved-questions]: #unresolved-questions
+
+- Which license should we use for RFCs in this repository?
+- How can we send out FCPs to all interested parties? (The mailing list would
```
```suggestion
- How can we send out FCP notifications to all interested parties? (The mailing list would
```

petervdonovan commented on text/0002-rfc-process.md:
```diff
@@ -0,0 +1,485 @@
+
+---
+
+Text to be placed in `0000_template.md`:
```
```suggestion
Text to be placed in `0000-template.md`:
```

lhstrh commented on 0000-template.md:
```diff
@@ -0,0 +1,70 @@
+
+- Clearly name and introduction new concepts.
+- Use diagrams to help illustrate your ideas where applicable.
+- Include LF code examples and use-cases if you propose changes to the language.
```
```suggestion
- Include LF code examples and use cases if you propose changes that are user facing.
```

lhstrh commented on 0000-template.md:
```diff
@@ -0,0 +1,70 @@
+- Clearly name and introduction new concepts.
+- Use diagrams to help illustrate your ideas where applicable.
+- Include LF code examples and use-cases if you propose changes to the language.
+- Also consider to use code snippets and examples if you are proposing an interface or
```
Covered by "user-facing changes" already.
```suggestion
```

lhstrh commented on 0000-template.md:
```diff
@@ -0,0 +1,70 @@
+- Use diagrams to help illustrate your ideas where applicable.
+- Include LF code examples and use-cases if you propose changes to the language.
+- Also consider to use code snippets and examples if you are proposing an interface or
+  system contract.
```
```suggestion
```

lhstrh commented on 0000-template.md:
```diff
@@ -0,0 +1,70 @@
+- Include LF code examples and use-cases if you propose changes to the language.
+- Also consider to use code snippets and examples if you are proposing an interface or
+  system contract.
+- Describe the interaction with other components and features.
```
```suggestion
- Describe the interaction with existing components and features.
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+[Lingua Franca RFCs]: #lf-rfcs
+
+The RFC (request for comments) process is a framework for managing both
+historical and planned changes within the LF community. The record of active
```
```suggestion
historical and planned changes to the Lingua Franca toolchain. The record of active
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+
+## TL;DR
+
+Follow these steps to open a new RFC:
```
```suggestion
To open a new RFC, follow these steps:
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+
+The RFC (request for comments) process is a framework for managing both
+historical and planned changes within the LF community. The record of active
+RFCs tells the story of our project, our technology and our thought process. We
```
```suggestion
RFCs tells the story of our project, our technology, and our thought process. We
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+1. Fork or clone this repository.
+2. Copy [`0000-template.md`](0000-template.md) to `rfcs/0000-my-feature.md`,
+   replacing `my-feature` with a descriptive short name for your proposal.
+3. Fill in the current date in the top section, but leave the RFC PR and
```
```suggestion
3. Enter the current date in the top section, but leave the RFC PR and
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+   replacing `my-feature` with a descriptive short name for your proposal.
+3. Fill in the current date in the top section, but leave the RFC PR and
+   tracking issue unchanged for now.
+4. Fill in the sections of the template to start your proposal and push your
```
```suggestion
4. Provide content for the sections of the template to start your proposal and push your
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+  enough to adapt to the concrete scope and impact of a proposal, and only
+  involve the stakeholders directly impacted by the community.
+- Specification or even standardization of language features and interfaces. If
+  we decide to create specifications for certain aspects of the language and the
```
```suggestion
  we decide to create specifications for certain aspects of the language or the
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+
+Opening a PR will allow the stakeholders and interested community members to
+review your proposal and raise concerns or make suggestions. We will also
+assign a [shepherd](#role-of-the-shepherd) to your PR who helps manage the
```
```suggestion
assign a [shepherd](#role-of-the-shepherd) to your PR to help manage the
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+Opening a PR will allow the stakeholders and interested community members to
+review your proposal and raise concerns or make suggestions. We will also
+assign a [shepherd](#role-of-the-shepherd) to your PR who helps manage the
+remaining process. At this stage, the shepherd should request reviews from
```
```suggestion
remaining process. A shepherd will request reviews from
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+to help reach an agreement. The shepherd may help to moderate the discussion if
+needed. The acceptance criteria provide a guideline on
+
+Make changes as new commits to the pull request, and leave a comment on the pull
```
Do we really want to insist on this? I think this comes at the risk of getting a messy history and would prefer people would be expected to use common sense.

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+days and is announced both on the community mailing list and the [`#rfc` Zulip
+channel](https://lf-lang.zulipchat.com/#narrow/stream/425418-RFCs). The
+announcement should include the RFC title and abstract as well as a link to the
+PR. At the next developer meeting after the FCP period ends, the maintainers
```
```suggestion
PR. After the FCP period ends, the maintainers
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+If the RFC is accepted, the shepherd announces the new active RFC on the
+community mailing list and in the [`#rfc` Zulip channel](https://lf-lang.zulipchat.com/#narrow/stream/425418-RFCs).
+
+Furthermore, the shepherd creates an issue (or issues) that tracks the
```
I would leave this as a recommendation or leave it out. This relies to much on discipline and is too costly if we try to enforce it.

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+as a pull request to the relevant repo. Being active, however, does not
+guarantee that an implementation of the feature will ultimately be merged; it
+does mean that, in principle, all the major stakeholders have agreed to the
+feature and are amenable to merging it.
```
```suggestion
feature would like to see it realized. Approval of the PR also depends on other factors like code quality, testing, etc.
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,249 @@
+## When you need to follow this process
+[When you need to follow this process]: #when-you-need-to-follow-this-process
+
+You need to follow this process if you intend to make any semantic or syntactic
```
I think we might be a bit too strict here. I can suggest a rewrite of this paragraph.

lhstrh reviewed (commented):
Made a pass and left some comments.

lhstrh commented on README.md:
```diff
@@ -1,2 +1,237 @@
+
+#### Final comments and decision
+
+After the discussion settles down, the shepherd sends out a final call for
```
```suggestion
After the discussion settles down, the shepherd announces the start of a final comment
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,237 @@
+#### Final comments and decision
+
+After the discussion settles down, the shepherd sends out a final call for
+comments (FCP) with a disposition to accept or reject the RFC. The FCP lasts 7
```
```suggestion
period (FCP) with a disposition to accept or reject the RFC. The FCP lasts 7
```

lhstrh commented on README.md:
```diff
@@ -1,2 +1,237 @@
+days and is announced both on the community mailing list and the [`#rfc` Zulip
+channel](https://lf-lang.zulipchat.com/#narrow/stream/425418-RFCs). The
+announcement should include the RFC title and abstract as well as a link to the
+PR. After the FCP period ends, the maintainers
```
```suggestion
PR. After the FCP ends, the maintainers
```

lhstrh reviewed (commented):
Proper usage of the FCP acronym.

oowekyala commented on README.md:
```diff
@@ -1,2 +1,237 @@
+pull request to ask for comments. The PR will be reviewed by the stakeholders
+and other interested parties. You help to address concerns and incorporate
+suggestions in the document. At some point, we will decide to merge the PR, in
+which case the RFC becomes active, or close the PR, in which case the RFC is
```
IIUC, the "active" state for a RFC means something like "accepted in principle by the community". But it says nothing of the implementation status of the feature, right? After the RFC has been implemented, does the RFC stay "active", or should it have another state, like "delivered"? Or is this state supposed to be captured by the state of the tracking issue?

cmnrd replied:
Good question. I would say the tracking issue captures this sufficiently. Probably we just have to see how this goes. It only really becomes a problem once we have a bunch of accepted RFCs, and we can always fine tune when we notice a problem.

oowekyala replied:
True, this will not be a problem from the start. Thanks for your reply

tanneberger commented on 0000-template.md:
```diff
@@ -0,0 +1,68 @@
+- Clearly name and introduction new concepts.
+- Use diagrams to help illustrate your ideas where applicable.
+- Include LF code examples and use cases if you propose changes that are user facing.
+- Describe the interaction with existing components and features.
```
add: - Specify Interfaces that ensure interoperability between components.
